Spec the processing model for the Web Audio API

padenot opened the pull request:
This lays down some infrastructure we can build upon.

The cycle detection and ordering is not finished, though.

joeberkovitz commented:
Paul, do you have an ETA on this? Some of the things to do with audio time / DOM time mapping are best done after this PR gets merged since I would like to express them in terms of rendering thread actions.

padenot commented:
I'm picking that up now, sorry for the delay.

joeberkovitz commented:
This is to be reviewed by 10 Dec.

joeberkovitz commented on index.html:
```diff
@@ -666,6 +678,45 @@ <h2 id="AudioContext">
           document.
         </p>
+        <p>
+          <!-- ReSpec does not support documenting constructor -->
```
Actually, ReSpec does support constructor docs although it's a bit weird. Use this approach inside a <dl class="idl"> element for the interface:
     <dt>Constructor</dt>
     <dd>
       <dl class="parameters">
        ...
    </dd>

joeberkovitz commented on index.html:
```diff
@@ -666,6 +678,45 @@ <h2 id="AudioContext">
+          execute these steps:</span>
+        </p>
+        <ol>
+          <li>Set a <code>control thread flag</code> to
```
Elsewhere you refer to this as a "control thread state" flag, should be consistent.

joeberkovitz commented on index.html:
```diff
@@ -1374,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+            <a>AudioContext</a> in a document is most of the time unnecessary.
+          </p>
+          <p>
+            Additionaly, user-agent can have an implementation-defined maximum
```
sp: Additionally

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+          performance.
+        </p>
+        <p>
+          Additionaly, the Worker model makes creating a dedicated thread
```
sp: Additionally

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+        <p>
+          The <dfn>control thread</dfn> is the thread from
+          which the <a>AudioContext</a> is instanciated, and from which authors
```
sp: instantiated

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        <p>
+          Each <a>AudioContext</a> has a single <dfn>control message
+          queue</dfn>, that is a list of <dfn data-lt="control message">control
+          messages</dfn>, that are a operations running on the <a>control
```
remove spurious "a"

billhofmann replied:
and comma after </dfn>

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+
+        <p>
+          <dfn id="queuing">Queuing an control message</dfn> means adding the
```
an -> a

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        <p>
+          The <dfn>control thread</dfn> is the thread from
+          which the <a>AudioContext</a> is instanciated, and from which authors
+          manipulate the audio graph. The <dfn>rendering
```
It would be helpful if there was more of a identification between the term _control thread_ and the thread from which an operation is being invoked. You do say the thread "...from which authors manipulate the audio graph" but would it be clearer to identify the control thread with the thread in which page-level scripts run?

padenot replied:
Yes.

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+          <em>Q<sub>rendering</sub></em>.
+          </li>
+          <li>
+          Remove the oldest element of <em>Q<sub>rendering</sub></em>.
```
I don't understand the repeated iterations through Qrendering and the nested "Until..." loops in this area. It seems as though there may be some sort of copy/paste error that has repeated some material? Confusingly, the first loop refers to _elements_ of the queue and the next loop refers to _messages_ in the queue... which makes it seem like there is a mixup somewhere here.

padenot replied:
Yes, sorry, fixed.

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+              <li>For each <a>AudioNode</a> <em>node</em> of <em>nodes</em>
+                <ol>
+                  <li>
+                    <em>node</em> is a <a>DelayNode</a> part of a cycle, add it
```
"If..." is missing here.

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+                  </li>
+                  <li>Mark <em>node</em>.
+                  </li>
+                  <li>Add <em>node</em> to the beginning of <em>ordered</em>.
```
This algorithm seems to result in nodes being ordered with the _destination_ end of the graph at the _beginning_ of the ordered list. That seems the reverse of what is intended, give the subsequent section that appears to be intended to work through the graph from sources towards destinations, making buffers available for reading by downstream nodes.

padenot replied:
I think I was missing a "ed" in connected. It walks from input nodes to input nodes, walking towards source nodes. It's not complete I think.

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+              </li>
+            </ol>
+          </li>
+          <li>For each <a>DelayNode</a> in a cycle, <a href="availabel">make
```
Mistyped href

joeberkovitz commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+              </li>
+              <li>If <em>nodes</em> contains cycles, mute all the
+              <a>AudioNode</a>s that are part of this cycle, and remove them
+              from <em>nodes</em>.
```
It really feels as though cycle detection could occur in conjunction with the node-visiting-and-marking scheme by maintaining a list of nodes already encountered during a given recursive node traversal; if any of these get visited, then a cycle has been found and that list will hold the contents of the cycle.

padenot replied:
I was trying to make things simple here, we don't have performance constraints.

billhofmann commented on index.html:
```diff
@@ -666,6 +678,45 @@ <h2 id="AudioContext">
+              attribute of the <a>AudioContext</a> to <code>"running"</code>.
+              </li>
+              <li>Queue a task to fire a simple event named
+              <code>"statechange"</code> at the <a>AudioContext</a>.
```
Minor: event names aren't quoted elsewhere

billhofmann commented on index.html:
```diff
@@ -783,56 +835,223 @@ <h2 id="AudioContext">
+              <em>promise</em>.
+              </li>
+              <li>If the <a href="#widl-AudioContext-state">state</a> attribute
+              of the <a>AudioContext</a> is already <code>"running"</code>,
```
remove "'s

billhofmann commented on index.html:
```diff
@@ -783,56 +835,223 @@ <h2 id="AudioContext">
+              return it, and abort these steps.
+              </li>
+              <li>If the <em>control thread state</em> flag on the
+              <a>AudioContext</a> is <code>"closed"</code> reject the promise
```
quotes?

billhofmann commented on index.html:
```diff
@@ -783,56 +835,223 @@ <h2 id="AudioContext">
+                  </li>
+                  <li>If the <a href="#widl-audiocontext-state">state</a>
+                  attribute of the <a>AudioContext</a> is not already
+                  <code>"closed"</code>:
```
quotes, here and elsewhere.  Should be generally consistent.

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+      <section>
+        <h3> Control thread and rendering thread </h3>
+        <p>
+          The Web Audio API MUST be implemented using a <a>control thread</a>,
```
Wording:  I'd suggest:
"The Web Audio API MUST be implemented using a <a>control thread</a> and a <a>rendering thread</a>."

padenot replied:
Yes, this was leftover from a previous iteration.

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+        <p>
+          <a>Control messages</a> in a <a>control message queue</a> are ordered by
+          date of insertion. The <dfn>oldest message</dfn> is therefore the one
```
time of insertion?

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+          "https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel">
+          in parallel</a>, as described in [[HTML]].
+        </p>
+        <p>
```
Is this a new section (e.g., <h3>)?  Or should we retitle the entire section? "Asynchronous and Synchronous operations"

padenot replied:
This describes asynchronous operations, that are implemented internally using a synchronous and an asynchronous part. I've tried to be a bit more clear about it.

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+          Rendering an audio graph
+        </h3>
+        <p>
+        Rendering an audio graph is done by blocks of 128 samples-frames.
```
by blocks --> in blocks

padenot replied:
No, sample-frame is the technical term in various specs. It means "the ensemble of sample that happen across multiple channels, at the same time". We use frame in Gecko to say the same thing, but yeah, I see the term sample-frame often in maths and spec text.

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+
+        <p>
+        Operations that happens <dfn data-lt="atomic">atomically</dfn> on a given thread, can only
```
happens --> happen
on a given thread, can --> on a given thread can

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+        <p>
+        The algorithm for rendering a block of audio from an <a>AudioContext</a>
+        <em>G</em> with an <a href=#control-message-queue>control message queue</a>
+        <em>Q</em> is as follow:
```
is as follows.

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+            href=#control-message-queue>control message queue</a>.
+          <a>Atomically</a> swap <em>Q<sub>rendering</sub></em> and
+          <em>Q</em></li>
+        <li> Until there is no more event in <em>Q<sub>rendering</sub></em>,
```
Until there are no more events in...

billhofmann commented on index.html:
```diff
@@ -6883,6 +7140,303 @@ <h2 id="MediaStreamAudioSourceNode">
+          queue</a>. <a>Atomically</a> <a>swap</a> <em>Q<sub>rendering</sub></em> and
+          <em>Q</em>
+          </li>
+          <li>Until there is no more messages in <em>Q<sub>rendering</sub></em>,
```
is --> are

cwilso commented on index.html:
```diff
@@ -100,6 +100,18 @@
+      content: "TODO:";
+    }
+    .synchronous:before {
+      content: "⌛ ";
```
Is this standard in any other specs?  If not, I'd use the word, not an emoji.

padenot replied:
Yes, this is what is used in HTML.

skratchdot commented on index.html:
```diff
@@ -666,6 +678,45 @@ <h2 id="AudioContext">
+          It is unfortunately not possible to programatically notify authors
+          that the creation of the <a>AudioContext</a> failed. User-Agents are
+          encouraged to log an informative message if they have access to a
+          logging mecanism, such an a developer tools console.
```
fix typo: "mecanism" to "mechanism"
fix typo: "such an a" to "such as a"

skratchdot commented on index.html:
```diff
@@ -783,56 +835,223 @@ <h2 id="AudioContext">
+              </li>
+            </ol>
+            <p class="note">
+              When an <a>AudioContext</a> is has been closed, implementation
```
fix typo: "is has" to "has"

skratchdot commented on index.html:
```diff
@@ -783,56 +835,223 @@ <h2 id="AudioContext">
+            </ol>
+            <p class="note">
+              When an <a>AudioContext</a> is has been closed, implementation
+              can choose to aggressively release more resource than when
```
fix type: "more resource" to "more resources"

skratchdot commented on index.html:
```diff
@@ -1374,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+          </h3>
+          <p>
+            <a>AudioContext</a>s should be considered expensive objects.
+            Creating an <a>AudioContext</a> often involves creating an
```
grammar: "creating an high-priority" to "creating a high-priority"

skratchdot commented on index.html:
```diff
@@ -1374,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+            Creating an <a>AudioContext</a> often involves creating an
+            high-priority thread, and use a low-latency system audio stream,
+            both having impact on energy consumption. Creating more than one
+            <a>AudioContext</a> in a document is most of the time unnecessary.
```
grammar: "creating an high-priority" to "creating a high-priority"
grammar: "and use a low-latency" to "and using a low-latency"
grammar: "both having impact on" to "both having an impact on"

skratchdot commented on index.html:
```diff
@@ -1374,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+            Constructing or resuming an <a>AudioContext</a> involves <dfn id=
+            "acquiring">acquiring system resources</dfn>. This means opening an
+            system audio stream. This operation returns when the the audios
+            stream is ready.
```
typo: "when the the audios stream" to "when the audio stream"

skratchdot commented on index.html:
```diff
@@ -1374,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+          <p>
+            Constructing or resuming an <a>AudioContext</a> involves <dfn id=
+            "acquiring">acquiring system resources</dfn>. This means opening an
+            system audio stream. This operation returns when the the audios
```
"opening an system" to "opening a system"

padenot commented:
This addresses all comments, in particular:
- Rework of the ordering
- Make it clear that the event loop thing are just for spec purposes, and describes the necessary guarantees
- Typos, quotes, grammar, etc.
- Rebase against master, tidy
- Suspend vs. pausing audio hardware

rtoy commented on index.html:
```diff
@@ -820,25 +832,28 @@ <h2 id="BaseAudioContext">
-            transition to "running", and then to "closed" once audio processing
-            has completed and <a><code>oncomplete</code></a> has been fired.
+            state MUST begin in <code>suspended</code>, and transitions to
+            running when system resources are acquired and audio has begun
```
Wrap running in <code>, like you did for suspended?

rtoy commented on index.html:
```diff
@@ -820,25 +832,28 @@ <h2 id="BaseAudioContext">
+            state MUST begin in <code>suspended</code>, and transitions to
+            running when system resources are acquired and audio has begun
+            processing. For OfflineAudioContexts, the state will remain in
+            <code>suspended</code> until <code>startRendering()</code> is
```
Use <a> for startRendering() instead of <code> so we get a link?

rtoy commented on index.html:
```diff
@@ -820,25 +832,28 @@ <h2 id="BaseAudioContext">
-              will cause a transition to "running", or a call to
-              <code>close()</code> will cause a transition to "closed".
+              When the state is <code>suspended</code>, a call to
+              <code>resume()</code> will cause a transition to
```
<a> for resume() to get a link?

rtoy commented on index.html:
```diff
@@ -820,25 +832,28 @@ <h2 id="BaseAudioContext">
-              <code>close()</code> will cause a transition to "closed".
+              When the state is <code>suspended</code>, a call to
+              <code>resume()</code> will cause a transition to
+              <code>running</code>, or a call to <code>close()</code> will
```
<a> for close()?

rtoy commented on index.html:
```diff
@@ -820,25 +832,28 @@ <h2 id="BaseAudioContext">
-              will cause a transition to "suspended", or a call to
-              <code>close()</code> will cause a transition to "closed".
+              When the state is <code>running</code>, a call to
+              <code>suspend()</code> will cause a transition to
```
<a> for suspend()?

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
-              <a>BaseAudioContext</a> for some time, and wishes to let the
-              audio hardware power down. The promise resolves when the frame
+              <a>BaseAudioContext</a> for some time, and wishes to temporarily
+              <a>release system resource</a> associated with te
```
Typo: "te" -> "the"

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
               immediately (with no other effect) if the context is already
-              suspended. The promise is rejected if the context has been
-              closed.
+              <code>"uspended"</code>. The promise is rejected if the context
```
Typo: "uspended" -> "suspended".  To match other changes we probably want to remove the quotes.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+            <ol>
+              <li>Let <em>promise</em> be a new Promise.
+              </li>
+              <li>If the <em>control thread state</em> flag on the
```
Is "control thread state" defined somewhere? While reading this in order, I didn't see it.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              with <code>InvalidStateError</code>, abort these steps, returning
+              <em>promise</em>.
+              </li>
+              <li>If the <a href="#widl-AudioContext-state">state</a> attribute
```
Is this state attribute intended to be the same as the "control thread state" above?  If yes, make it consistent.

padenot replied:
No. AudioContext.state is what is visible from content. The "control thread state" is an internal state.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              <em>promise</em>.
+              </li>
+              <li>If the <a href="#widl-AudioContext-state">state</a> attribute
+              of the <a>AudioContext</a> is already <code>suspended"</code>,
```
Typo: Remove quote after suspended.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              thread</a>:
             </p>
+            <ol>
+              <li>Attempt to <a>release system resources</a>.
```
Chrome doesn't release system resources; the audio system is paused.

padenot replied:
I'll make it so that it's required to release resources when calling `close()`, and that implementation can choose to release some resources when doing `suspend()`. Tracing some code in chrome, I see that there is at least a thread being released.

padenot replied:
I'll note though that the current prose requires it:

> "and then allows the system to release its claim on audio hardware"

I was merely factoring this into its own sectino.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              window functions, the data is considered as a continuous stream -
+              i.e. the <code>resume()</code>/<code>suspend()</code> does not
+              cause silence to appear in the <a>AnalyserNode</a>'s stream of
+              data.
```
I think the behavior of the analyser node needs to be expanded on.  If the context is suspended and I keep calling analyser node, do I get exactly the same data until the context is resumed? I think that makes the most sense

padenot replied:
I agree.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              with <code>InvalidStateError</code>, abort these steps, returning
+              <em>promise</em>.
+              </li>
+              <li>If the <a href="#widl-AudioContext-state">state</a> attribute
```
How exactly is this item different from the item at line 1039?

padenot replied:
This is to prevent a race across the control thread and the rendering thread when doing, say:

```
context.close();
context.suspend();
```

When calling suspend, the state is not "closed", because it's updated when the audio system has been released, per spec. We still want to reject the `suspend` call by throwing, so we use an internal flag for this.

rtoy commented on index.html:
```diff
@@ -853,66 +868,227 @@ <h2 id="BaseAudioContext">
+              </li>
+            </ol>
+            <p>
+              Running a <a>control message</a> to suspend an
```
Typo?  "suspend" -> "close"?

rtoy commented on index.html:
```diff
@@ -1431,6 +1593,44 @@ <h3 id="lifetime-AudioContext" class="informative">
+            <a href="#widl-AudioContext-resume-Promise-void">close</a> allow
+            authors to <dfn id="releasing">release system resources</dfn>.
+            <a href="#releasing">Releasing system resources</a> means releasing
+            the system resources such as threads, processes, audio streams, but
```
Is this required?  Chrome only releases threads and streams when closing.  Suspending just pauses the audio stream.

rtoy commented on index.html:
```diff
@@ -1445,6 +1645,50 @@ <h2 id="AudioContext">
+              execute these steps:</span>
+            </p>
+            <ol>
+              <li>Set a <code>control thread state</code> to
```
You used <em>control thread state</em> in other places.  Be consistent.

padenot replied:
This is the flag in question, being initialized in the ctor. Or maybe I don't understand your comment ?

rtoy replied:
Yeah, I updated the comment.  basically you used `<em>` in other places but used `<code>` here.

rtoy commented on index.html:
```diff
@@ -1445,6 +1645,50 @@ <h2 id="AudioContext">
+            <ol>
+              <li>Attempt to <a href="#acquiring">Acquire system resources</a>.
+              </li>
+              <li>In case of failure, abort these steps.
```
What happens if we can't acquire the system resources? No error thrown?

padenot replied:
See the `<p class="note">` below, we can't throw here because opening an audio channel is async, but UA can warn the user.

Maybe Chrome can throw because it has an hard-coded limit ?

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+          Background
+        </h3>
+        <p>
+          Real-time audio systems that require low latency are, often time,
```
", often time," -> "often"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+        <p>
+          For this reason, the traditional way of executing asynchronous
+          operations on the Web Platform, the event loop, does not work here,
+          as the thread is no <em>continuously executing</em>. Additionally, a
```
Typo: "no" -> "not"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+          performance.
+        </p>
+        <p>
+          Additionally, the Worker model makes creating a dedicated thread
```
Is this dedicated thread different from the audio thread where AudioNodes run?

padenot replied:
Yes, Workers are on their own thread, separate from the control and rendering threads.

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+          queue</a> of this <a>rendering thread</a>.
+        </p>
+        <p>
+          The synchronous and asynchronous sections order with respect to other
```
What?  Can you simplify/explain this sentence a bit more?

padenot replied:
Well it just means you can't have async and sync sections reordered, I can add that.

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+        <p>
+          The synchronous and asynchronous sections order with respect to other
+          events MUST be the same: given two operation <em>A</em> and
```
Why em here? You used var for the queue variables above.

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+        <p>
+          The synchronous and asynchronous sections order with respect to other
+          events MUST be the same: given two operation <em>A</em> and
+          <em>B</em> with respectively synchronous and asynchronous section
```
typo: "respectively" -> "respective"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+        </p>
+        <p>
+          The algorithm for rendering a block of audio from an
+          <a>AudioContext</a> <em>G</em> with an <a href=
```
Typo: "an" -> "a"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+          queue</a>. <a>Atomically</a> <a>swap</a>
+          <em>Q<sub>rendering</sub></em> and <em>Q</em>
+          </li>
+          <li>Until there are no more messages in
```
I think it reads better to say while there are messages in Qrendering instead until there are no messages.

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+              It will contain an ordered list of <a>AudioNode</a>s when this
+              ordering algorithm terminates.
+              </li>
+              <li>Let <em>nodes</em> be the set of all nodes belonging to an
```
You haven't defined what this set really means.  Is it the set of all nodes created on the AudioContext?

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+              </li>
+              <li>For each <a>AudioNode</a> <em>node</em> in <em>nodes</em>
+                <ol>
+                  <li>If <em>node</em> is a <a>DelayNode</a> part of a cycle,
```
"DelayNode part of a cycle" -> "DelayNode that is part of a cycle"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+              </li>
+              <li>While there are unmarked <a>AudioNode</a>s in <em>nodes</em>:
+                <ol>
+                  <li>Choose an <a>AudioNode</a> <em>node</em> in
```
Choose how?

padenot replied:
It does not matter I think, this is a topological sort.

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+              </li>
+            </ol>
+          </li>
+          <li>Inverse the order of <em>nodes</em>.
```
"Inverse" -> "Reverse"

rtoy commented on index.html:
```diff
@@ -7176,6 +7414,334 @@ <h2 id="MediaStreamAudioSourceNode">
+                  buffer</dfn>.
+                  </li>
+                  <li>
+                    <a href="#computation-of-value">Compute the value</a> of
```
"value" -> "values" since it could be a-rate.

rtoy commented:
I'm mostly happy with the changes, once the typos are fixed up.  I'd prefer to see this merged sooner rather than later, and then fix up any issues we discover. (It's easier to read and understand in context that a diff.)

padenot commented:
@rtoy, I've addressed your comments in the last commit..

padenot commented:
@rtoy, only the last four commits are now compared to what you looked at. They are what was in other pull requests, but PRs based on PRs was messy so I closed the other one.

This is not a lot of changes, and should be the last things needed.

padenot commented:
Comments addressed.

rtoy commented:
Except for the one typo, lgtm.

padenot commented:
I'm merging this, @rtoy has gave an LGTM, and it has become so big that it's a pain to rebase.

Onto `AudioWorklet`s now.

rtoy commented:
Yeah, it will be much easier to read and understand this in its entirety as part of the spec.  If we messed up something, we can fix it later.
